rust/lance-linalg/src/kernels.rs

```diff
@@ -47,24 +58,54 @@ where
     max_idx
 }
 
-/// Argmin on a [PrimitiveArray].
+/// Argmin on an iterator. Fused the operation in iterator to avoid memory allocation.
 ///
 /// Returns the index of the min value in the array.
-pub fn argmin<T: ArrowNumericType>(array: &PrimitiveArray<T>) -> Option<u32>
+///
+pub fn argmin<T: Num + PartialOrd + Copy>(iter: impl Iterator<Item = T>) -> Option<u32>
 where
-    T::Native: PartialOrd + Bounded,
+    T: Bounded,
 {
+    argmin_value(iter).map(|(idx, _)| idx)
+}
+
+pub fn argmin_value<T: Num + Bounded + PartialOrd + Copy>(
```

Comment:
this function seems to be the same as `argmin_value_opt `?

Reply:
One takes `Iterator<Item=Option<T>>` the other takes `Iterator<Item=T>`.   Arrow's primitive array `Float32Array::iter()` is a `Iterator<Option<T>>`.

Reply:
Can probably just let `argmin_val` to call `argmin_val_opt` with little performance penalty tho.

Reply:
oh, i was expected the opt is in the return argument. ok this makes sense.